gadget/gadget.go

```diff
@@ -445,16 +445,35 @@ func readInfo(f func(string) ([]byte, error), gadgetYamlFn string, model Model)
 
 // ReadInfo reads the gadget specific metadata from meta/gadget.yaml in the snap
 // root directory.
-func ReadInfo(gadgetSnapRootDir string, model Model) (*Info, error) {
+// If validationConstraints is not nil, it also performs consistency rules validation as Validate does using the given validationConstraints.
```

Comment:
I see the distinction here now about nil vs an empty object but it still feels a bit confusing to me as a reader of the code that the behavior for this function when `ReadInfo(...,nil)` is not the same as `ReadInfo(...,&gadget.ValidationConstraints{})`... perhaps I need to be better at reading things more closely, but it does feel like a mistake that is too easy to make where someone intends to not do special extra validation like around the encrypted data partition et al, but still wants default validation to happen and thinks that the default behavior of nil == &gadget.ValidationConstraints is the same when in fact it is not.

Maybe instead we could do something like 

```go
type ValidationConstraints struct{
    SkipBasicValidation bool
    ...
}
```

I normally would be more okay with this but given the recent problems we have had around doing the unintended validation on gadgets, I think more verbosity around validation is a net good thing

Reply:
but that approach would exactly produce the problem that you would get validation by default, which in our situations is actually what we don't want. I don't see a non-confusing solution to this that doesn't involve adding yet one more argument. 

My reasoning was that passing  a nil for validationConstraints would give you no validation, it seems reasonable enough to me.

Reply:
perhaps then we should do the inverse of what I initially proposed and need something like 
```go
type ValidationConstraints struct{
    BasicVolumeValidation bool
    ...
}
```

such that `&gadget.ValidationConstraints{}` still behaves the same as `nil`, but for the cases where we want basic validation of the volumes it would then read as

```go
validateConstraints := &gadget.ValidationConstraints{
    BasicVolumeValidation: true,
}
... := gadget.ReadInfo(..., validateConstraints)
```

where now it is clear to the reader what validation is being performed and with something like:

```
... := gadget.ReadInfo(..., &gadget.ValidationConstraints{})
```
it is more obvious that no validation is happening since the empty struct is being passed

Reply:
I tried to address this, it's a bit more complicated because ValidationConstraints where passed to both ReadInfo and Validate with slightly different behavior, so I had to really invent add a new type. Overall it's still a bit awkward but the cleanest path (just dropping the 3rd ReadInfo param) has its own concerns about maybe forgetting to call Validate when needed.

Reply:
stared a bit at my last changes and looked at the usage patterns again, also compared/constrated with what we do with snaps:  it seems we have a use case where we want ReadInfoFromSnapFile without validation, that will only delay errors but will move them to a more obvious place, that case is sysconfig.ConfigureTargetSystem. It seems that the cleanest if a bit redundant approach is two have 4 functions:

ReadInfo
ReadInfoAndValidate
ReadInfoFromSnapFile
ReadInfoFromSnapFileNoValidate

this capture the preferred/common patterns, and support the non-common ones with longer names. For completness their docs can point always as the other variant.

Reply:
I think your newest proposal makes the most sense to me of all our current options

Reply:
done